### PR TITLE
Shave ~90K off regex in native aot

### PR DESCRIPTION
**Pull request by stephentoub**
- ifdef out some code only used in debug but that included a static which was rooting some of the methods
- Avoid using a ValueTuple`7 as a dictionary key and instead use a custom struct
- Remove a Sort call that was leftover from debugging but isn't necessary for correctness or performance
- Avoid using a `HashSet<UnicodeCategory>` when a `bool[30]` would suffice
- Avoid a generic instantation for a `Dictionary<enum,object,object>` by changing it to `Dictionary<int,object,object>`, which we already use in the same type

This:
```C#
using System.Text.RegularExpressions;

internal class Program
{
    static void Main()
    {
        Console.WriteLine(new Regex("a", RegexOptions.CultureInvariant | RegexOptions.IgnoreCase).IsMatch("abc"));
    }
}
```
published for Windows x64 native aot drops from 3618K  to 3529K.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs [hunk 1]
```diff
@@ -36,7 +36,7 @@ internal sealed class CharSetSolver : ISolver<BDD>
         /// This cache is necessary for the recursive operation algorithms to be guaranteed linear time.
         /// A well-crafted character class could otherwise cause execution time to be exponential.
         /// </remarks>
-        private readonly Dictionary<(BooleanOperation op, BDD a, BDD? b), BDD> _operationCache = new();
+        private readonly Dictionary<(int op, BDD a, BDD? b), BDD> _operationCache = new(); // op is BooleanOperation; using int to reuse generic instantiation
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs:39]: (nit)

```suggestion
        private readonly Dictionary<(int op, BDD a, BDD? b), BDD> _operationCache = new(); // op is BooleanOperation; using int to reuse generic instantiation with _bddCache above
```

Might help explain a little bit more.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs [hunk 2]
```diff
@@ -418,11 +418,18 @@ BDD Compute(string set)
...
 
-                    // Collect individual category codes
-                    var categoryCodes = new HashSet<UnicodeCategory>();
+                    // Collect individual category codes. At this point, the only codes that can be
+                    // here are valid UnicodeCategories; SpaceConst is never used as part of groups,
+                    // as such groups are only constructed in our own RegexCharClass consts, and it's
+                    // not used there.
+                    const int UnicodeCategoryCount = 1 + (int)UnicodeCategory.OtherNotAssigned;
+                    Debug.Assert(!Enum.IsDefined((UnicodeCategory)UnicodeCategoryCount));
+                    var categoryCodes = new bool[UnicodeCategoryCount];
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:427]: Would it make sense to `stackalloc` this (outside of the loop of course) so we aren't allocating an array in a loop?

**Review by eerhardt — APPROVED** [review]
This is great!

LGTM. Thanks

**Comment by EgorBo** [comment]
> Avoid using a ValueTuple`7 as a dictionary key and instead use a custom struct

Sounds like https://github.com/dotnet/runtime/issues/76513?

**Pull request merged**